cmd/gpu-scheduling-webhook/main.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bombsimon/logrusr/v3"
+	"github.com/go-logr/logr"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const nvidiaGPU = "nvidia.com/gpu"
+
+var (
+	nvidiaGPUToleration = corev1.Toleration{
+		Key:      nvidiaGPU,
+		Operator: corev1.TolerationOpEqual,
+		Value:    "true",
+		Effect:   corev1.TaintEffectNoSchedule,
+	}
+
+	opts = options{}
+
+	rootCmd = &cobra.Command{
+		Use:   "gpu-scheduling-webhook",
+		Short: "Controls where pods will be scheduled when they request a GPU",
+		Long: `Controls where pods will be scheduled when they request a GPU.
+
+Example:
+$ gpu-scheduling-webhook --cert-dir=<cert-dir> --port=443`,
+		RunE: RunE,
+	}
+)
+
+func init() {
+	rootCmd.Flags().StringVar(&opts.certDir, "cert-dir", "", "A folder holding the server private key and and certicate for TLS")
+	rootCmd.Flags().IntVar(&opts.port, "port", 0, "Port the server will listen on")
+}
+
+func setupLogger() logr.Logger {
+	innerLogger := logrus.New()
+	innerLogger.Formatter = &logrus.JSONFormatter{}
+	log.SetLogger(logrusr.New(innerLogger))
+	return log.Log
+}
+
+type options struct {
+	certDir string
+	port    int
+}
+
+type gpuTolerator struct{}
+
+func (*gpuTolerator) Default(ctx context.Context, obj runtime.Object) error {
+	logger := log.FromContext(ctx)
+
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return fmt.Errorf("expected a Pod but got a %T", obj)
+	}
+
+	logger = logger.WithValues("pod", fmt.Sprintf("%s/%s", pod.Namespace, pod.Name))
+
+	// Check if a GPU required on resources from containers and init containers.
+	var gpuNeeded bool
+	gpuNeeded = hasNvidaGPURequest(logger, pod.Spec.InitContainers)
+
+	if !gpuNeeded {
+		gpuNeeded = hasNvidaGPURequest(logger, pod.Spec.Containers)
+	}
+
+	if gpuNeeded {
```

Comment:
maybe something like
```
if hasNvidaGPURequest(logger, pod.Spec.InitContainers) || hasNvidaGPURequest(logger, pod.Spec.Containers)
```
?

Reply:
I'd like to save some iterations and skip the second loop if the first one has already found the request.